hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/FieldReaderAppender.java

```diff
@@ -91,31 +93,34 @@ public Size apply(
 		if ( !dispatcherType.isPrimitive() ) {
 			methodVisitor.visitTypeInsn( Opcodes.CHECKCAST, persistentFieldAsDefined.getType().asErasure().getInternalName() );
 		}
-		fieldWrite( methodVisitor );
+		fieldWrite( methodVisitor, persistentFieldAsDefined );
 		// end if
 		methodVisitor.visitLabel( skip );
 		if ( implementationContext.getClassFileVersion().isAtLeast( ClassFileVersion.JAVA_V6 ) ) {
 			methodVisitor.visitFrame( Opcodes.F_SAME, 0, null, 0, null );
 		}
 		// return field
 		methodVisitor.visitVarInsn( Opcodes.ALOAD, 0 );
-		fieldRead( methodVisitor );
+		fieldRead( methodVisitor, persistentFieldAsDefined );
+		if ( !persistentField.getType().isPrimitive() ) {
```

Comment:
Maybe rather `if ( persistentField.getType().asErasure().getInternalName().equals( persistentFieldAsDefined.getType().asErasure().getInternalName() ) )`?
Though I suppose it's no big deal if we introduce an unnecessary cast. Not sure if compiler optimizations are applied to this generated code.